File: 224/python/main.py
class Solution:
    def calculate(self, s: str) -> int:
        def compute(value, sign, n, operand):
            value[n] += operand * sign[n]
            sign[n] = 1

        value = [0] * 512
        sign = [1] * 512
        found_operator = True
        n = 0
        
        len_s = len(s)
        i = 0
        while i < len_s:
            char = s[i]
            
            if char.isspace():
                None

            elif char == '(':
                n += 1
                value[n] = 0
                sign[n] = 1
                found_operator = True
                
            elif char == ')':
                n -= 1
                compute(value, sign, n, value[n+1])
                found_operator = False

            elif found_operator == False:
                if char == '-':
                    sign[n] *= -1
                found_operator = True

            elif found_operator and not char.isdigit():
                if char == '-':
                    sign[n] *= -1
            
            elif char.isdigit():
                digits = char
                while i + 1 < len_s and s[i+1].isdigit():
                    i += 1
                    digits += s[i]
                operand = int(digits)
                compute(value, sign, n, operand)
                found_operator = False
            
            i += 1
        
        return value[0]

File: 224/c/main.c
#include <ctype.h>
#include <stdbool.h>

int calculate(char* s) {
    
    int values[512] = { 0 };
    bool signs[512] = { false };

    int i = 0;
    int n = 0;
    bool seek_operator = false;

    int value = 0;
    int sign = true;
    char c;
    while ( (c = s[i]) ) {

        if (isspace(c)) {}
        else if (c == '(') {
            values[n] = value;
            signs[n] = sign;
            sign = true;
            value = 0;
            n++;
            seek_operator = false;
        } else if (c == ')') {
            n--;
            if (!signs[n]) {
                value *= -1;
            }
            value += values[n];
            sign = true;
            seek_operator = true;
        } else if (seek_operator) {
            if (c == '-') {
                sign = false;
            }
            seek_operator = false;
        } else if (!isdigit(c)) {
            if (c == '-') {
                sign = !sign;
            }
        } else {
            int operand = 0;
            do {
                operand = operand * 10 + (c - '0');
                c = s[++i];
            } while ( isdigit(c) );
            i--;
            if (!sign) {
                operand *= -1;
            }
            value += operand;
            sign = true;
            seek_operator = true;
        }
        i++;
    }

    return value;
}
